man/man1/ado_pl.1
.TH ado_pl 1 "25 May 1998"
.SH NAME
ado_pl \- simple Adonis pipeline 
.SH SYNOPSIS
.B ado_pl
.B [options]
.SH DESCRIPTION
.LP
.B ado_pl
is an Adonis specific command. It handles all basics of sky-subtraction,
flat-field division and bad pixel correction. It needs one or several
data cubes containing both sky and object planes, and optionally a
flat-field and a bad pixel map.
.B ado_pl
is very specific to Adonis in the way it recognizes how data are
organized in the file. It can be used with other data, however,
provided they are separated into an object and a sky file.
.LP
The 2 supported data formats are:
.LP
Data has been acquired by offseting the telescope. Sky and object
are provided in separated files. This is the SEPARATED functioning
mode, which algorithm is described below. Use the -a and -s
options to declare the object and sky file names.
.LP
Chopping has been done, with whatever acquisition pattern. This
is the PACKED functioning mode, algorithm described below.
Use the -p option to declare the packed cube name.
.LP
If acquisition does not match any of these schemes, make the
best use of
.B extract
and 
.B catcube
to reformat data into an acceptable format.
.LP 
Once the sky has been correctly subtracted, the user may want
to flat-field the result and/or correct the bad pixels. These
calibration data must have been prepared before, and are provided
through the -f (flat-field) and -b (bad pixel map) options on
command line.
.LP 
As usual, a flat-field is a linear gain map which mean value is
1.0, it shows the pixel-to-pixel response of the detector.
A bad pixel map is in standard eclipse format: an 8-bit FITS file,
with pixel value of 1 for good pixels and 0 for bad ones.
.SH ALGORITHM
.LP
1. SEPARATED DATA
.LP
The algorithm is:
.LP
* Average the sky to sky_avg
.br
* Subtract sky_avg from each plane in the object cube
.br
* Average the sky-subtracted object if requested.
.br
* Append results to the output cube
.LP
2. PACKED DATA
.LP
Data is acquired in cycles of object/sky. The algorithm is:
.LP
* Extract every cycle in the cube. For each cycle:
.br
* Extract all sky planes in this cycle
.br
* Average the sky planes to one: sky_avg
.br
* Subtract sky_avg from all object planes in cycle
.br
* Average the resulting sky-subtracted object planes if requested
.br
* Append results to the output cube
.LP
The output will have one plane per cycle in the initial cube. No average
is made on the output, for it may need some shift-and-add, or selection, or
wiser averaging. Make use of other eclipse tools to reduce this cube to
one image.
.SH OPTIONS
.TP
.B \-p " filename"
to declare a packed cube name, in PACKED data mode only.
.TP
.B \-a " filename"
to declare an object filename, in SEPARATED data mode only. 
.TP
.B \-s " filename"
to declare a sky filename, in SEPARATED data mode only. 
.TP
.B \-o " filename"
to declare an output filename, any mode. Default is calib.fits.
.TP
.B \-f " filename"
to declare a flat-field file name to use with this data. The output
planes will all be divided by this image. PACKED or SEPARATED mode.
.TP
.B \-b " filename"
to declare a bad pixel map file name to use with this data. The output
planes will all be bad pixel corrected using this bad pixel map.
The default behaviour is: look for a file named 'badpixmap' in current
directory, which fits the usual eclipse format for pixel maps, and use
it. If this file is not available or the -b option was not used, do
not correct for bad pixels.
PACKED or SEPARATED mode.
.TP
.B \-1
1 (one) requests an average of object frames to a single frame. In the
case of SEPARATED data, it is equivalent to averaging all object
frames to a single one. In the case of PACKED data, it is equivalent to
outputting one plane per cube cycle.
.SH EXAMPLE
.LP
For a PACKED input cube named f0118.fits, with the following cycle
keywords in its FITS header:
.br
NAXIS3 = 240 
.br
OJ_N_SEQ = 30
.br
OJ_N_IMA = 4
.br
OB_CYCL = 01
.LP
Requesting a processing with option -1 active will create one average
object frame per cycle, i.e. 30 frames. Without this option on, it will
create one object frame per input object frame, i.e. 120.
.SH FILES
.LP
Input files shall all comply with FITS format.
.SH SEE ALSO
.LP
.B arcube, cstcube, average, flat, norm, deadpix, shiftadd
.SH AUTHOR
.LP
N. Devillard -- ESO


CHANGES
List of changes between eclipse 4.9 and eclipse 5.0
---------------------------------------------------

Since last delivery, there has been no major channge in the library.
Only a couple of minor fixes were brought to the ISAAC pipeline.



ins/adonis/recipes/ado_pl.c

/*---------------------------------------------------------------------------
	
    File name 	:	ado_pl.c
	Author 		:	Nicolas Devillard
	Created on	:	Sept 2, 1996
	Description	:	Basic Adonis data reduction pipeline

 *--------------------------------------------------------------------------*/

/*

 $Id: ado_pl.c,v 1.3 2001/01/24 13:43:50 ndevilla Exp $
 $Author: ndevilla $
 $Date: 2001/01/24 13:43:50 $
 $Revision: 1.3 $

 */

/*---------------------------------------------------------------------------
 * 								Includes
 *--------------------------------------------------------------------------*/

#include <stdio.h>
#include <string.h>
#include "eclipse.h"
#include "ado_utils.h"

static void usage(char *pname) ;
static char prog_desc[] = "simple Adonis pipeline" ;

 
/*---------------------------------------------------------------------------
 * 								Main
 *--------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
	char		object[FILENAMESZ],
				sky[FILENAMESZ],
				flat[FILENAMESZ],
				bpm[FILENAMESZ],
				output[FILENAMESZ] ;
	int			c ;
	int			mode = CALIBRATION_UNKNOWN ;
	int			flag_avg = 0 ;

	object[0] = sky[0] = flat[0] = (char)0 ;
	(void)strcpy(bpm, "badpixmap") ;
	(void)strcpy(output, "calib.fits") ;
	
    /* 
	 * Command line parsing by getopt()
	 * See man page for getopt(3c) and implementation in ./pipes
	 */
    while ((c = getopt(argc, argv, "LvDM:a:p:o:s:f:b:1")) != EOF)
        switch(c)
        {
		/* Standard option: display license (not documented in usage)	*/
			case 'L':
			eclipse_display_license() ;
			return 0 ;

		/* Cube name for packed (chopped) cubes */
			case 'p':
			(void)strcpy(object, optarg) ;
			mode = CALIBRATION_PACKED ;
			break ;

		/* Cube name for object cubes	*/
			case 'a':
			(void)strcpy(object, optarg) ;
			mode = CALIBRATION_SEPARATED ;
			break ;

		/* Output name	*/
			case 'o':
			(void)strcpy(output, optarg) ;
			break ;

		/* Cube name for sky cube */
			case 's':
			(void)strcpy(sky, optarg) ;
			mode = CALIBRATION_SEPARATED ;
			break ;

		/* Cube name for flat cube */
			case 'f':
			(void)strcpy(flat, optarg) ;
			break ;

		/* Bad pixel map name	*/
			case 'b':
			(void)strcpy(bpm, optarg) ;
			break ;

		/* Averaging flag */
			case '1':
			flag_avg = 1 ;
			break ;

            default:
            usage(argv[0]) ;
            break ;
        }

	/* Initialize eclipse environment */
	eclipse_init();

	switch (mode) {
		case CALIBRATION_PACKED:
		reduce_packed_cube(object, flat, bpm, output, flag_avg) ;
		break ;

		case CALIBRATION_SEPARATED:
		reduce_separated_cube(object, sky, flat, bpm, output, flag_avg) ;
		break ;

		default:
		e_error("unkown calibration scheme: aborting") ;
		break ;
	}

	if (debug_active())
		xmemory_status() ;
	return(0) ;
}

/*
 * This function only gives the usage for the program
 */

static void
usage(char *pname)
{
	hello_world(pname, prog_desc) ;
	printf("use : %s\n", pname) ;
	printf("\t[-p filename] to give a packed cube name\n") ;
	printf("\t[-a filename] to give an object cube name\n") ;
	printf("\t[-o filename] to give an output file name\n") ;
	printf("\t[-s filename] to give a sky cube name\n") ;
	printf("\t[-f filename] to give a gain map name\n") ;
	printf("\t[-b filename] to give a bad pixel map name") ;
	printf(" (default: badpixmap)\n") ;
	printf("\t[-1] to request an average object for each cycle\n") ;
	printf("\n\n") ;
	exit(0) ;
}
